src/yunohost/certificate.py

```diff
@@ -0,0 +1,751 @@
+                                          CA=CERTIFICATION_AUTHORITY)
+
+    intermediate_certificate = requests.get(INTERMEDIATE_CERTIFICATE_URL).text
+
+    # Now save the key and signed certificate
+    logger.info("Saving the key and signed certificate...")
+
+    # Create corresponding directory
+    date_tag = datetime.now().strftime("%Y%m%d.%H%M%S")
+
+    new_cert_folder = "%s/%s-history/%s-letsencrypt" % (CERT_FOLDER, domain, date_tag)
+    os.makedirs(new_cert_folder)
+
+    _set_permissions(new_cert_folder, "root", "root", 0655)
+
+    # Move the private key
+    shutil.move(domain_key_file, os.path.join(new_cert_folder, "key.pem"))
+
+    # Write the cert
+    domain_cert_file = os.path.join(new_cert_folder, "crt.pem")
+
+    with open(domain_cert_file, "w") as f:
+        f.write(signed_certificate)
+        f.write(intermediate_certificate)
+
+    _set_permissions(domain_cert_file, "root", "metronome", 0640)
+
+    _enable_certificate(domain, new_cert_folder)
+
+    # Check the status of the certificate is now good
+    status_summary = _get_status(domain)["summary"]
+
+    if status_summary["code"] != "great":
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_certificate_fetching_or_enabling_failed', domain=domain))
+
+
+def _prepare_certificate_signing_request(domain, key_file, output_folder):
+    # Init a request
+    csr = crypto.X509Req()
+
+    # Set the domain
+    csr.get_subject().CN = domain
+
+    # Set the key
+    with open(key_file, 'rt') as f:
+        key = crypto.load_privatekey(crypto.FILETYPE_PEM, f.read())
+
+    csr.set_pubkey(key)
+
+    # Sign the request
+    csr.sign(key, "sha256")
+
+    # Save the request in tmp folder
+    csr_file = output_folder + domain + ".csr"
+    logger.info("Saving to %s.", csr_file)
+
+    with open(csr_file, "w") as f:
+        f.write(crypto.dump_certificate_request(crypto.FILETYPE_PEM, csr))
+
+
+def _get_status(domain):
+    cert_file = os.path.join(CERT_FOLDER, domain, "crt.pem")
+
+    if not os.path.isfile(cert_file):
+        return {}
+
+    try:
+        cert = crypto.load_certificate(crypto.FILETYPE_PEM, open(cert_file).read())
+    except Exception as exception:
+        import traceback
+        traceback.print_exc(file=sys.stdout)
+        raise MoulinetteError(errno.EINVAL, m18n.n('certmanager_cannot_read_cert', domain=domain, file=cert_file, reason=exception))
+
+    cert_subject = cert.get_subject().CN
+    cert_issuer = cert.get_issuer().CN
+    valid_up_to = datetime.strptime(cert.get_notAfter(), "%Y%m%d%H%M%SZ")
+    days_remaining = (valid_up_to - datetime.now()).days
+
+    if cert_issuer == _name_self_CA():
+        CA_type = {
+            "code": "self-signed",
+            "verbose": "Self-signed",
+        }
+
+    elif cert_issuer.startswith("Let's Encrypt"):
+        CA_type = {
+            "code": "lets-encrypt",
+            "verbose": "Let's Encrypt",
+        }
+
+    elif cert_issuer.startswith("Fake LE"):
```

Comment:
Not sure why I added this. I think it was mainly intended for the case where we would like to have, for instance a --fake option (or a debug flag) to test the install using the Fake/Staging Let's Encrypt CA instead of the production one, to be able to run the code multiple times without the fear of the rate limit (20 certificate per period of 7 days per subdomains IIRC). This could also be used before an actual install to test that the ACME challenge is working... Any thoughts on this ?